[CDAP-21172] Implement Basic Mutation Operations for Metadata Tables

123-komal opened the pull request:
### **Description:**
This PR is the follow-up for #15971 and here we are temporarily building this module separately as a workaround for a dependency issue that breaks the unified build. It's key changes are listed below:

- **SpannerMetadataStorage**: In this PR basic Metadata operations and `batch` method  has been implemented along with its helper methods.This change is responsible for the mutation execution in the spanner metadata tables.

### **Testing:**

- **Unit Test**: Test cases have been  added for verifying the `createIndex` and `readFromSpanner` functionality, ensuring that both behave as expected in different scenarios.
-  **Local Testing**:Tested these changes using a Distributed docker image.

bhardwaj-priyanshu commented on cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java:
```diff
@@ -74,6 +75,8 @@
 import javax.annotation.Nullable;
 import javax.net.ssl.SSLContext;
+
+import io.cdap.cdap.spi.metadata.VersionedMetadata;
```
is import order correct?

123-komal replied:
corrected

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+        return intermediary.getChange();
+      });
+    } catch (SpannerException e) {
+      if (e.getErrorCode() == com.google.cloud.spanner.ErrorCode.ABORTED) {
```
Do we need the if else here? Why not directly throw IOException here?

123-komal replied:
okk updated

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+                               MutationOptions options) throws IOException {
+    try {
+      for (Mutation mutation : mutations) { // Iterate over the list of mutations
+        if (mutation.getOperation() == Mutation.Op.INSERT || mutation.getOperation() == Mutation.Op.UPDATE ||
```
Why not make a variable of `mutation.getOperation()` and use it everywhere?

123-komal replied:
updated

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+          mutation.getOperation() == Mutation.Op.INSERT_OR_UPDATE) {
+          transaction.buffer(mutation);
+        } else if (mutation.getOperation() == Mutation.Op.DELETE) {
+          transaction.buffer(mutation);
```
why do we need else here?

We are performing the same step : `transaction.buffer(mutation);` in both cases.

You can also make another method `validateMutation` and check if type belongs to list of one of the supported enums.

123-komal replied:
removed

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+        }
+      }
+    } catch (SpannerException e) {
+      if (e.getErrorCode() == ErrorCode.ALREADY_EXISTS) {
```
do we need to check this code? Exception will already have necessary details.

123-komal replied:
updated

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+        Struct row = resultSet.getCurrentRowAsStruct();
+
+        // Get the JSON string from the Spanner JSON column
+        String metadataString = row.getJson(0); // Convert Spanner JSON to String
```
instead of index here, you can use the field name constant itself.

123-komal replied:
okk

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+    boolean duplicate = false;
+
+    for (MetadataMutation mutation : mutations) {
+      if (!entities.add(mutation.getEntity())) {
+        duplicate = true;
+        break;
+      }
+      mutationMap.put(mutation.getEntity(), mutation);
+    }
+
+    if (duplicate) {
+      // If there are multiple mutations for the same entity, execute all in sequence
+      List<MetadataChange> changes = new ArrayList<>(mutations.size());
+      for (MetadataMutation mutation : mutations) {
+        changes.add(apply(mutation, options));
+      }
+      return changes;
+    }
```
can be structured in a better way.

123-komal replied:
updated

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +307,271 @@ public void dropIndex() throws IOException {
+   * @return the list of all changes performed by the mutations, if all mutations complete
+   * successfully
+   */
+  private List<MetadataChange> doSpannerBatch(LinkedHashMap<MetadataEntity, MetadataMutation> mutations
```
can be renamed bit better?

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/pom.xml:
```diff
@@ -55,6 +55,22 @@
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-common</artifactId>
+      <version>6.12.0-SNAPSHOT</version>
```
there should be some variable for this?

123-komal replied:
okay

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/pom.xml:
```diff
@@ -55,6 +55,22 @@
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
```
why this is getting added two times?

123-komal replied:
updated

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/pom.xml:
```diff
@@ -55,6 +55,22 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
```
test dependencies should be mentioned in separate section

123-komal replied:
okay

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerChangeRequest.java:
```diff
@@ -0,0 +1,51 @@
+    this.change = change;
+  }
+
+  public List<Mutation> getMutation() {
```
add java docs?

123-komal replied:
in the constructor Java docs we explained the mutation its just a getter for the same do I need to add the repeative Java doc for the same too?

sidhdirenge replied:
Yes - it is recommended all public methods should have the javadocs.  The PR will have checkstyle warnings in such cases.

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +92,11 @@ public void initialize(MetadataStorageContext context) throws Exception {
     this.databaseId = Objects.requireNonNull(properties.get("database"));
 
-    this.spanner = SpannerOptions.newBuilder().setProjectId(projectId).build().getService();
+    String emulatorHost = Objects.requireNonNull(properties.get("SPANNER_EMULATOR_HOST"));
```
will this be non-null in case of running in production? i thought emulator is very specifc to unit-tests?

sidhdirenge replied:
+1

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+   * @return existing metadata along with its version in the table, or an empty metadata with null
+   * version.
+   */
+  public VersionedMetadata readFromSpanner(MetadataEntity entity, ReadContext transaction) throws IOException {
```
why this is public? are we using it anywhere else?

sidhdirenge replied:
If it is used in test, make it default (no access identifier specified) and the add annotation `@VisibleForTesting`

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+
+  /**
+   * Translate a metadata entity into a metadata_id in the table.
+   * TODO (CDAP-21172): MetadataUtil currently requires the cdap-common dependency.
```
this is not causing guava conflict?

123-komal replied:
no

sidhdirenge replied:
please remove unnecessary TODOs.

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+   * TODO (CDAP-21172): MetadataUtil currently requires the cdap-common dependency.
+   * TODO (CDAP-21172): We need to refactor this to remove the dependency.
+   */
+  protected String toMetadataId(MetadataEntity entity) {
```
this function can be simplified I think. It seems to be a little hacky to me.

sidhdirenge replied:
If you are reusing existing method from ES - please move it to the util class.

123-komal replied:
updated

sidhdirenge commented on cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/ChangeRequest.java:
```diff
@@ -0,0 +1,35 @@
+ * This interface defines a command that encapsulates the necessary information
+ * for a specific storage provider to execute a metadata modification.
+ */
+public interface ChangeRequest {
```
This ChangeRequest is not externally used. In spanner / elastic specific implementations, they use their own form of change requests.

https://github.com/search?q=repo%3Acdapio%2Fcdap+RequestAndChange&type=code

No need of making an interface for this.

sidhdirenge commented on cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ESChangeRequest.java:
```diff
@@ -16,19 +16,20 @@
  * change that it effects.
  */
-public class RequestAndChange {
+public class ESChangeRequest implements ChangeRequest {
```
Separate interface impl not needed

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerChangeRequest.java:
```diff
@@ -0,0 +1,51 @@
+
+/**
+ * A {@link ChangeRequest} implementation backed by Google Cloud Spanner.
+ */
```
Separate interface impl not needed

123-komal replied:
updated

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+   * @throws IOException              if a Spanner error occurs during buffering.
+   * @throws IllegalArgumentException if any mutation has an unsupported operation type.
+   */
+  private void executeMutation(TransactionContext transaction, MetadataEntity entity, List<Mutation> mutations)
```
why do we need to pass entity? just to throw exception?

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+   *                                  (INSERT, UPDATE, INSERT_OR_UPDATE, DELETE).
+   */
+  private void validateMutationOperation(Mutation.Op operation) {
+    if (!(operation == Mutation.Op.INSERT
```
can we have a static variable of the set of supported mutations?

and here we just check if the set contains that op or not?

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -306,3 +583,24 @@ public synchronized void close() {
   }
 }
 
+class MetadataTable {
```
We can move this to separate class file. No need to make it an inner class here.

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+  public VersionedMetadata readFromSpanner(MetadataEntity entity, ReadContext transaction) throws IOException {
+    try {
+      Statement statement = Statement.newBuilder(
+          "SELECT metadata_column, version FROM metadata "
```
you can use the string constants here instead of hardcoding the field names.

123-komal replied:
okk

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+        return remove(before, (MetadataMutation.Remove) mutation);
+      default:
+        throw new IllegalStateException(String.format("Unknown mutation type '%s' for %s", mutation.getType(),
+                                                      mutation));
```
since there is no implementation for create, drop update and all its very hard to understand if there can be any improvement possible in the logic here.

Can we just have applyMutation as unimplemented and not have all these function in this PR

@sidhdirenge  wdyt?

sidhdirenge replied:
SGTM. Okay to keep it unimplemented in this PR.

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+
+        Metadata metadata = gson.fromJson(metadataString, Metadata.class);
+
+        return VersionedMetadata.of(metadata, version);
```
nit - no need of extra line breaks here.

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+  public List<Mutation> writeToSpanner(MetadataEntity entity, Long expectVersion, Metadata metadata) throws
+    IOException {
     throw new IOException("NOT IMPLEMENTED");
   }
```
please remove this if not needed as mentioned in the above comment

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+    LinkedHashMap<MetadataEntity, MetadataMutation> mutationMap = new LinkedHashMap<>(mutations.size());
+
+    if (containsDuplicates(mutations, mutationMap)) {
+      // If there are multiple mutations for the same entity, execute all in sequence
```
nit - `execute all in sequence` means you need to just execute the mutations corresponding to an entity in sequence.

However, you are planning to execute all mutations in sequence in case of duplicates as the mutations might be linked.

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+
+    LinkedHashMap<MetadataEntity, MetadataMutation> mutationMap = new LinkedHashMap<>(mutations.size());
+
+    if (containsDuplicates(mutations, mutationMap)) {
```
its confusing here. The mutaionMap is changed by containsDuplicates and then again reused by doBatch?

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+    }
+
+    // Spanner Batch Processing (No Duplicates)
+    return doBatch(mutationMap);
```
I am not sure but what is the benifit of doing batch in case of spanner? applying mutations I assume will not require any network call hence fery fast?

sidhdirenge commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,394 @@
+    spannermetadatastorage.createIndex();
+  }
+
+  private static final class MockMetadataStorageContext implements MetadataStorageContext {
```
can be moved to end of file.

sidhdirenge replied:
This is a private inner class - needs to be moved to end of file.

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +286,282 @@ public void dropIndex() throws IOException {
+    MetadataEntity entity = mutation.getEntity();
+    try {
+      TransactionRunner runner = dbClient.readWriteTransaction();
+      return runner.run(transaction -> {
```
Dont create new transaction for each apply its not efficient

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +101,11 @@ public void initialize(MetadataStorageContext context) throws Exception {
 
-    this.spanner = SpannerOptions.newBuilder().setProjectId(projectId).build().getService();
+    String emulatorHost = properties.get("SPANNER_EMULATOR_HOST");
+    this.spanner = SpannerOptions.newBuilder().setEmulatorHost(emulatorHost).setProjectId(projectId)
```
should you `setEmulatorHost` if the string is null or empty?

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +295,199 @@ public void dropIndex() throws IOException {
+   * @param mutations A list of {@link MetadataMutation} objects to apply. Cannot be null.
+   * @param options   {@link MutationOptions} for the batch operation.
+   * @return A {@link List} of {@link MetadataChange} objects for all successfully processed mutations.
+   * Returns an empty list if the input mutations list is empty or all mutations are skipped.
```
Please fix these checkstyle warnings.

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerChangeRequest.java:
```diff
@@ -0,0 +1,50 @@
+    return mutations;
+  }
+
+  public MetadataChange getChange() {
```
java docs for public method

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +102,15 @@ public void initialize(MetadataStorageContext context) throws Exception {
 
-    this.spanner = SpannerOptions.newBuilder().setProjectId(projectId).build().getService();
+    String emulatorHost = properties.get("SPANNER_EMULATOR_HOST");
+    SpannerOptions.Builder builder = SpannerOptions.newBuilder().setProjectId(projectId);
```
nit - rename to something better like `spannerOptionsBuilder` or `optionsBuilder`

123-komal replied:
okk

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -17,66 +17,81 @@
 public class SpannerMetadataStorage implements MetadataStorage {
 
   private static final Logger LOG = LoggerFactory.getLogger(SpannerMetadataStorage.class);
+  static final Gson gson = new GsonBuilder()
```
nit - rename to `GSON` as it is static final (constant)

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +300,184 @@ public void dropIndex() throws IOException {
+   * Create a mutation list {@link Mutation} for adding or updating the metadata for an entity in the
+   * table.
+   */
+  private List<Mutation> writeToSpanner(MetadataEntity entity, Long expectVersion, Metadata metadata) throws
```
This method is not actually performing any write. It is just creating a mutation for the given entity. Write will happen at the end of transaction right?

Better to rename the method than creating confusion that we are making multiple write calls to spanner (not very efficient) instead of clubbing them.

sidhdirenge commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -278,13 +300,184 @@ public void dropIndex() throws IOException {
+        MetadataEntity entity = mutation.getEntity();
+        // Read the existing metadata within the transaction.
+        VersionedMetadata existingMetadata = readFromSpanner(entity, transaction);
+        Metadata updatedMetadata;
```
can be moved to line # 450

123-komal replied:
okay

sidhdirenge commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,394 @@
+      throw new RuntimeException("An unexpected error occurred during Spanner read transaction.", e);
+    }
+
+    // Verify Test Case 1: Existing Metadata
```
It is recommended in unit testing for each test method to focus on a single, specific test case.

You can separate out these test cases into 2 different methods : `testReadFromSpanner_existingEntity()`, `testReadFromSpanner_nonExistentEntity()`

Commit:
[CDAP-21172] Implement Basic Mutation Operations for Metadata Tables

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

updated

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +300,193 @@ public void dropIndex() throws IOException {
+    if (!keyValuePairs.isEmpty()) {
+      builder.append(":").append(keyValuePairs);
+    }
+
```
```
protected String toMetadataId(MetadataEntity entity) {
    final boolean isEntityTypeVersioned = MetadataUtil.isVersionedEntityType(entity.getType());

    String keyValuePairs = StreamSupport.stream(entity.spliterator(), false)
        .filter(kv -> !isEntityTypeVersioned || !MetadataEntity.VERSION.equalsIgnoreCase(kv.getKey()))
        .map(kv -> kv.getKey() + "=" + kv.getValue())
        .collect(Collectors.joining(","));

    return keyValuePairs.isEmpty() ? entity.getType() : entity.getType() + ":" + keyValuePairs;
}
```

Can we write something like this?

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,380 @@
+      assertTrue("TextNgramIndex should exist", metadataTableIndexes.contains("TextNgramIndex"));
+
+      assertTrue("ValueNgramIndex should exist", metadataPropsTableIndexes.contains("ValueNgramIndex"));
+
```
remove extra lines

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,380 @@
+      assertTrue("UserNgramIndex should exist", metadataTableIndexes.contains("UserNgramIndex"));
+      assertTrue("SystemNgramIndex should exist", metadataTableIndexes.contains("SystemNgramIndex"));
+      assertTrue("TextNgramIndex should exist", metadataTableIndexes.contains("TextNgramIndex"));
+
```
remove extra lines

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataTables.java:
```diff
@@ -0,0 +1,41 @@
+    public static final String NESTED_SCOPE_FIELD = "scope";
+    public static final String NESTED_VALUE_FIELD = "value";
+  }
+
```
remove extra space

bhardwaj-priyanshu commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataTables.java:
```diff
@@ -0,0 +1,41 @@
+
+package io.cdap.cdap.metadata.spanner;
+
+public class SpannerMetadataTables {
```
SpannerMetadataTables.MetadataTable

doesnt this sounds redundant?

wdyt @sidhdirenge

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerChangeRequest.java:
```diff
@@ -0,0 +1,61 @@
+ * A simple class to pass around a list of Spanner mutations, along with the
+ * metadata change that it effects.
+ */
+public class SpannerChangeRequest {
```
nit: this can be renamed to `ChangeRequest`. Since this is in `io.cdap.cdap.metadata.spanner` package, `Spanner` is redundant in the naming.

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+    throws IOException{
+    MetadataEntity entity = mutation.getEntity();
+    VersionedMetadata before = readFromSpanner(entity, transaction);
+    assert before != null;
```
Use `Preconditions.checkArgument(...)` instead of `assert`

123-komal replied:
okk

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+   * @return The resulting {@link MetadataChange}.
+   */
+  private MetadataChange applyAndBufferMutation(TransactionContext transaction, MetadataMutation mutation)
+    throws IOException{
```
nit: missing whitespace before `{`

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+   * @param options   {@link MutationOptions} for the batch operation.
+   * @return A {@link List} of {@link MetadataChange} objects for all successfully processed mutations.
+   * @throws IOException If a Spanner error occurs or an underlying operation fails.
+   * TODO (CDAP-21172): Mutation Execution can further be optimized.
```
nit: please fix checkstyle warnings

Fix t/o

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+   * @return existing metadata along with its version in the table, or an empty metadata with null version.
+   */
+  @VisibleForTesting
+  VersionedMetadata readFromSpanner(MetadataEntity entity, ReadContext transaction) throws
```
nit: use better naming to describe the purpose of the function

Can be renamed as `getVersionedMetadata()` or `readVersionedMetadata()`

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -17,66 +17,81 @@
+  static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(ScopedName.class, new ScopedNameTypeAdapter()).create();
+  private static final Set<Mutation.Op> SUPPORTED_MUTATION_OPS =
+    Collections.unmodifiableSet(new HashSet<>(java.util.Arrays.asList(
```
`import java.util.Arrays;`

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -17,66 +17,81 @@
 import io.cdap.cdap.spi.metadata.MutationOptions;
 import io.cdap.cdap.spi.metadata.Read;
+import io.cdap.cdap.spi.metadata.ScopedName;
+import io.cdap.cdap.spi.metadata.ScopedNameOfKind;
```
nit: unused imports can be removed 

Fix t/o

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -17,66 +17,81 @@
+import io.cdap.cdap.spi.metadata.VersionedMetadata;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import java.io.IOException;
```
Fix checkstyle warnings t/o

123-komal replied:
fixed

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -276,15 +298,157 @@ public void dropIndex() throws IOException {
+
+      ResultSet resultSet = transaction.executeQuery(statement);
+
+      if (resultSet.next()) {
```
Just curious, can there be more than one row in the resultSet?

123-komal replied:
no as the metadata_id is unique so we will have just one row.

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/Tables.java:
```diff
@@ -0,0 +1,40 @@
+
+package io.cdap.cdap.metadata.spanner;
+
+public class Tables {
```
Please add java docs.

Fix t/o

123-komal replied:
fixed

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,370 @@
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import java.io.IOException;
```
Fix checkstyle warnings t/o

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,370 @@
+                                   .setDisplayName("Test Instance")
+                                   .build()).get();
+    } catch (ExecutionException e) {
+      // Instance already exists, proceed.
```
Why is this being assumed?

vsethi09 replied:
Can we check the cause of the ExecutionException to confirm that instance already exists?

123-komal replied:
okk

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,370 @@
+    try {
+      dbAdminClient.createDatabase(INSTANCE_ID, DATABASE_ID, Collections.emptyList()).get();
+    } catch (ExecutionException e) {
+      // DB already exists, proceed.
```
Why is this being assumed?

vsethi09 replied:
Can we check the cause of the ExecutionException to confirm that DB already exists?

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,370 @@
+    configs.put("SPANNER_EMULATOR_HOST", EMULATOR_HOST);
+    MetadataStorageContext context = new MockMetadataStorageContext(configs);
+
+    spannermetadatastorage = new SpannerMetadataStorage();
```
nit: fix camel case in `spannermetadatastorage`

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,370 @@
+  }
+
+  @Before
+  public void setupForEachTest() throws IOException {
```
nit: simply`before()` or `beforeTest()`

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/Tables.java:
```diff
@@ -0,0 +1,43 @@
+ * Defines Spanner table and column name constants for managing metadata.
+ */
+public class Tables {
+  static class Metadata {
```
Nested classes can be public. Please add javadocs for public classes.

123-komal replied:
updated

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -16,67 +16,83 @@
 public class SpannerMetadataStorage implements MetadataStorage {
 
   private static final Logger LOG = LoggerFactory.getLogger(SpannerMetadataStorage.class);
+  static final Gson GSON = new GsonBuilder()
```
nit: As best practice please mention access modifier explicitly. I think it will be `private` in this case.

123-komal replied:
done

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +103,15 @@ public void initialize(MetadataStorageContext context) throws Exception {
     this.databaseId = Objects.requireNonNull(properties.get("database"));
 
-    this.spanner = SpannerOptions.newBuilder().setProjectId(projectId).build().getService();
+    String emulatorHost = properties.get("SPANNER_EMULATOR_HOST");
```
nit: Please add an explicit comment this property is used for unit tests.

Can CDAP sandbox use spanner emulator? If so, you can mention CDAP sandbox as well in the comment.

123-komal replied:
added

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +103,15 @@ public void initialize(MetadataStorageContext context) throws Exception {
     this.databaseId = Objects.requireNonNull(properties.get("database"));
 
-    this.spanner = SpannerOptions.newBuilder().setProjectId(projectId).build().getService();
+    String emulatorHost = properties.get("SPANNER_EMULATOR_HOST");
```
nit: rename property name to `emulator.host` or something better for consistency

System env variable name could be `SPANNER_EMULATOR_HOST`, but CDAP config properties don't follow the same naming convention.

123-komal replied:
fixed

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,369 @@
+          .set(Tables.Metadata.USER_FIELD).to(metadata.getTags().stream()
+                                                .filter(s -> false).map(ScopedName::getName)
+                                                .collect(Collectors.joining(",")))
+          .set(Tables.Metadata.SYSTEM_FIELD).to(metadata.getTags().stream().
```
nit: fix checkstyle warnings

123-komal replied:
fixed

vsethi09 commented on cdap-metadata-ext-spanner/src/test/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorageTest.java:
```diff
@@ -0,0 +1,369 @@
+        return null; // The callable must return something, null is fine for side effects
+      });
+    } catch (Exception e) {
+      throw new IOException("Failed to setup initial metadata for test: " + e.getMessage(), e);
```
nit: No need to use both `e.getMessage` for message and `e` for cause.

Simply: throw new IOException("Failed to setup initial metadata", e);

123-komal replied:
okk

vsethi09 commented on cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java:
```diff
@@ -87,8 +103,19 @@ public void initialize(MetadataStorageContext context) throws Exception {
+      does not persist data across sessions and is not suitable for production use.
+     */
+    String emulatorHost = properties.get("emulator.host");
+    SpannerOptions.Builder optionsbuilder = SpannerOptions.newBuilder().setProjectId(projectId);
```
nit: `optionsBuilder` instead of `optionsbuilder`
